debezium-core/src/main/java/io/debezium/cloudevents/CloudEventsMaker.java

```diff
@@ -0,0 +1,309 @@
+            case "postgresql":
+                return new PostgresCloudEventsMaker(parser);
+            case "mongodb":
+                return new MongodbCloudEventsMaker(parser);
+            case "sqlserver":
+                return new SqlserverCloudEventsMaker(parser);
+            default:
+                throw new DataException("No usable CloudEvents converters for connector type \"" + parser.connectorType() + "\"");
+        }
+    }
+
+    /**
+     * Create a concrete CloudEvents maker using the outputs of a record parser. Also need to specify the data content
+     * type (that is the serialization format of the data attribute) and the url of data schema registry when using Avro
+     * as the data content type.
+     *
+     * @param parser the parser of a change record
+     * @param contentType the data content type of CloudEvents
+     * @param schemaUrl the url of data schema registry; may be null
+     * @return a concrete CloudEvents maker
+     */
+    public static CloudEventsMaker create(RecordParser parser, SerializerType contentType, String schemaUrl) {
+        dataSchemaUrl = schemaUrl;
+        return create(parser, contentType);
+    }
+
+    CloudEventsMaker(RecordParser parser) {
+        recordParser = parser;
+    }
+
+    /**
+     * Construct the id of CloudEvents envelope.
+     *
+     * @return the id of CloudEvents envelope
+     */
+    public abstract String ceId();
+
+    /**
+     * Construct the source field of CloudEvents envelope.
+     *
+     * @return the source field of CloudEvents envelope
+     */
+    public String ceSource() {
+        return "/debezium/" + recordParser.connectorType();
+    }
+
+    /**
+     * Get the version of CloudEvents specification.
+     *
+     * @return the version of CloudEvents specification
+     */
+    public String ceSpecversion() {
+        return CLOUDEVENTS_SPECVERSION;
+    }
+
+    /**
+     * Construct the type field of CloudEvents envelope.
+     *
+     * @return the type field of CloudEvents envelope
+     */
+    public String ceType() {
+        return "io.debezium." + recordParser.connectorType() + ".datachangeevent";
+    }
+
+    /**
+     * Get the data content type of CloudEvents envelope.
+     *
+     * @return the data content type of CloudEvents envelope
+     */
+    public String ceDatacontenttype() {
+        return CONTENT_TYPE_NAME_MAP.get(dataContentType);
+    }
+
+    /**
+     * Get the data schema url of CloudEvents envelope.
+     *
+     * @return the data schema url of CloudEvents envelope
+     */
+    public String ceDataschema() {
+        return dataSchemaUrl;
+    }
+
+    /**
+     * Get the time of CloudEvents envelope.
+     *
+     * @return the time of CloudEvents envelope
+     */
+    public String ceTime() {
+        return recordParser.getMetadata(AbstractSourceInfo.TIMESTAMP_KEY).toString();
+    }
+
+    /**
+     * Construct the schema of the extrainfo field of CloudEvents envelope.
```

Comment:
what is `extrainfo` ?   Would it perhaps make sense to provide some CloudEvent extension for "extra info" stuff ?

Reply:
It's custom attributes defined by Debezium, such as the operation type ("create", "update" etc.), table name etc. See the discussion above, I already suggested to get rid of the "extrainfo" container and have them at the top-level.

> provide some CloudEvent extension

Can you remind me, how does that work again? How would it look different from the structure suggested so far?

Reply:
@matzew, do you have any reference to read on the "extension" concept? Thanks!